src/v2.rs

```diff
@@ -653,16 +656,54 @@ impl EszipV2 {
           wakers.push(cx.waker().clone());
           Poll::Pending
         }
-        EszipV2SourceSlot::Ready(bytes) => Poll::Ready(bytes.clone()),
+        EszipV2SourceSlot::Ready(bytes) => Poll::Ready(Some(bytes.clone())),
+        EszipV2SourceSlot::Taken => Poll::Ready(None),
       }
     })
     .await
   }
 
+  pub(crate) async fn take_module_source<'a>(
+    &'a self,
+    specifier: &str,
+  ) -> Option<Arc<Vec<u8>>> {
+    let source = poll_fn(|cx| {
+      let mut modules = self.modules.lock().unwrap();
+      let module = modules.get_mut(specifier).unwrap();
+      let slot = match module {
+        EszipV2Module::Module { source, .. } => source,
+        EszipV2Module::Redirect { .. } => {
+          panic!("redirects are already resolved")
+        }
+      };
+      match slot {
+        EszipV2SourceSlot::Pending { wakers, .. } => {
+          wakers.push(cx.waker().clone());
+          Poll::Pending
+        }
+        EszipV2SourceSlot::Ready(bytes) => Poll::Ready(Some(bytes.clone())),
+        EszipV2SourceSlot::Taken => Poll::Ready(None),
+      }
+    })
+    .await;
+    // Drop the source from memory.
+    let mut modules = self.modules.lock().unwrap();
+    let module = modules.get_mut(specifier).unwrap();
+    match module {
+      EszipV2Module::Module { source, .. } => {
+        *source = EszipV2SourceSlot::Taken;
+      }
+      EszipV2Module::Redirect { .. } => {
+        panic!("redirects are already resolved")
+      }
+    };
```

Comment:
```suggestion
      match slot {
        EszipV2SourceSlot::Pending { wakers, .. } => {
          wakers.push(cx.waker().clone());
          return Poll::Pending;
        }
        EszipV2SourceSlot::Ready(_) => {},
        EszipV2SourceSlot::Taken => return Poll::Ready(None),
      };
      let EszipV2SourceSlot::Ready(bytes) = std::mem::replace(source, EszipV2SourceSlot::Taken) else { unreachable!() };
      Poll::Ready(Some(bytes))
    })
    .await;
```

You can remove a lock.